File: 期末论文/1.2.2JavaScript实现异步IO/nodejs-soluction/asyncIO.js
const fs = require('fs');
const process = require('process');
let filePath = process.cwd() + '\\passage.txt';

//console.log(filePath);

fs.open(filePath,'r',(err,fd)=>{
    if(err){
        if(err.code === 'ENOENT'){
            console.log(`File:${filePath} doesn't exist!`);
        }
        else{
            console.log("error:",err);
        }
        return false;
    }
});

let buf;
//异步读
console.log("Start reading");
console.time("Time for read");
fs.readFile(filePath,(err,data)=>{
    if(err){
        console.log("error:",err);
        return false;
    }
    else{
        buf = data;
        console.timeEnd('Time for read');
        console.log(`End reading, words read: ${buf.length}`);
    }
});

//计算斐波那契数列
console.log("Start computing");
console.time("Time for computing");
let result = fn(44);
console.timeEnd("Time for computing");
console.log(`Finish computing, the result is ${result}`);


function fn(n){
    if(n === 0) return 0;
    if(n === 1) return 1;

    return fn(n-1)+fn(n-2);
}

File: 期末论文/1.2.1Java实现异步IO/src/asyncIO.java
import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousFileChannel;
import java.nio.channels.CompletionHandler;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.Future;
import java.nio.charset.Charset;

    public class asyncIO{
        public static void main(String[] args) throws Exception{
            asyncRead("src/passage.txt");
            System.out.println("Start computing");
            long start1 = System.currentTimeMillis();
            int result1 = fn(44);
            long end1 = System.currentTimeMillis();
            System.out.println("Finish computing, the result is " + result1 + " and it takes " + (end1 - start1) + "ms");

        }

        //异步io
        public static void asyncRead(String path) throws Exception{
            System.out.println("Start reading");
            long start = System.currentTimeMillis();

            Path file = Paths.get(path);
            AsynchronousFileChannel channel = AsynchronousFileChannel.open(file);

            ByteBuffer buffer = ByteBuffer.allocate(100000);
            // 回调式
            channel.read(buffer, 0, buffer,new CompletionHandler<Integer, ByteBuffer>() {
                @Override
                public void completed(Integer result, ByteBuffer attachment) {
                    long end = System.currentTimeMillis();
                    //反转buffer
                    buffer.flip();
                    //System.out.println(Charset.forName("UTF-8").decode(attchment).toString());
                    System.out.println("Finish reading, words read " + result + " and it takes " + (end - start) + "ms");
                }

                @Override
                public void failed(Throwable exc, ByteBuffer attachment) {
                    System.out.println("read error");
                }
            });
        }

        //斐波那契数列
        public static int fn(int n){
            if(n == 0) return 0;
            if(n == 1) return 1;

            return fn(n-1)+fn(n-2);
        }
    }


